Dockerfile

```diff
@@ -1,18 +1,62 @@
-FROM python:3.8-alpine
-LABEL maintainer="Jacob Tomlinson <jacob@tomlinson.email>"
+FROM python:3.9.6-alpine3.14 as builder
+LABEL maintainer="Rémy Greinhofer <remy.greinhofer@gmail.com>"
 
 WORKDIR /usr/src/app
-ARG EXTRAS=.[all,connector_matrix_e2e]
+
+ARG EXTRAS=[all,connector_matrix_e2e]
+ENV DEPS_DIR=/usr/src/app/deps
 
 # Copy source
 COPY . .
 
+# Install build tools and libraries to build OpsDroid and its dependencies.
 RUN apk update \
-    && apk add --no-cache build-base linux-headers musl-dev python3-dev libzmq zeromq-dev git openssh-client olm olm-dev libffi-dev \
-    && pip3 install --upgrade pip \
-    && pip3 install --no-cache-dir $EXTRAS \
-    && apk del build-base linux-headers musl-dev python3-dev zeromq-dev
+  && apk add \
+  build-base \
+  gcc \
+  g++ \
+  libffi-dev \
+  linux-headers \
+  musl-dev \
+  olm-dev \
+  openssh-client \
+  openssl-dev \
+  python3-dev \
+  zeromq-dev \
+  && pip install --upgrade \
+  build \
+  pip \
+  setuptools \
+  setuptools-scm \
+  wheel \
+  && mkdir -p "${DEPS_DIR}" \
+  && pip download --prefer-binary -d ${DEPS_DIR} .${EXTRAS} \
+  && pip wheel -w ${DEPS_DIR} ${DEPS_DIR}/*.tar.gz \
+  && python -m build --wheel --outdir ${DEPS_DIR}
+
+FROM python:3.9.6-alpine3.14 as runtime
+LABEL maintainer="Rémy Greinhofer <remy.greinhofer@gmail.com>"
+
+WORKDIR /usr/src/app
+
+ARG EXTRAS=[all,connector_matrix_e2e]
+ENV DEPS_DIR=/usr/src/app/deps
+
+# Copy the pre-built dependencies.
+COPY --from=builder ${DEPS_DIR}/*.whl ${DEPS_DIR}/
+
+# Install Opsdroid using only pre-built dependencies.
+RUN apk add --no-cache \
+  git \
+  libzmq \
+  && pip install --no-cache-dir --no-index -f ${DEPS_DIR} \
+  ${DEPS_DIR}/opsdroid-0+unknown-py3-none-any.whl${EXTRAS} \
```

Comment:
This version number looks funny. Is versioneer not working properly still?

Reply:
I think it is just because there is no git repository within the container. 

If I run it locally I get something like`opsdroid-0.23.0+1.g6d81c23-py3-none-any.whl`.

Reply:
Sure, but we want to fix this (see #950).

Switching to multi-stage builds here is a nice opportunity to resolve this problem because you can pull in the whole repo into the build stage and get the correct version in the package.

Reply:
This is not really relevant for the docker build though, as the package name as no impact on what is built. What would the benefit be? 

#950 seems to be a non-issue and should be closed.

Reply:
I disagree. The `opsdroid.__version__` variable is also set by versioneer at build time and is used in multiple places. #950 mainly covers ensuring that variable is correct.

In the docker images this gets set incorrectly currently, hence my request to fix it in this PR as part of the awesome work you're doing here.